Goal: Build a simple recommender system to recommend the best movies to users

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set()

In [2]:
metadata = pd.read_csv('movies_metadata.csv', low_memory=False)
pd.set_option('display.max_columns', None)
metadata.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943   
1  When siblings Judy and Peter discover an encha...  17.015539   
2  A family wedding reignites the ancient feud be...    11.7129   
3  Cheated on, mistreated and stepped on, the wom...   3.859495   
4  Just when George Banks has recovered from his ...   8.387519   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video  vote_average  vote_count  
0                    Toy Story  False           7.7      5415.0  
1                      Jumanji  False           6.9      2413.0  
2             Grumpier Old Men  False           6.5        92.0  
3            Waiting to Exh

In [3]:
metadata.shape

(45466, 24)

In [4]:
metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [5]:
metadata.rename(columns = {'vote_count' : 'V_ratings_count', 'vote_average' : 'R_average_ratings'}, inplace=True)

In [6]:
metadata.describe()

revenue       runtime  R_average_ratings  V_ratings_count
count  4.546000e+04  45203.000000       45460.000000     45460.000000
mean   1.120935e+07     94.128199           5.618207       109.897338
std    6.433225e+07     38.407810           1.924216       491.310374
min    0.000000e+00      0.000000           0.000000         0.000000
25%    0.000000e+00     85.000000           5.000000         3.000000
50%    0.000000e+00     95.000000           6.000000        10.000000
75%    0.000000e+00    107.000000           6.800000        34.000000
max    2.787965e+09   1256.000000          10.000000     14075.000000

In [7]:
metadata['release_date'].unique()

array(['1995-10-30', '1995-12-15', '1995-12-22', ..., '1904-03-05',
       '1991-05-13', '1917-10-21'], dtype=object)

In [8]:
metadata.count()

adult                    45466
belongs_to_collection     4494
budget                   45466
genres                   45466
homepage                  7782
id                       45466
imdb_id                  45449
original_language        45455
original_title           45466
overview                 44512
popularity               45461
poster_path              45080
production_companies     45463
production_countries     45463
release_date             45379
revenue                  45460
runtime                  45203
spoken_languages         45460
status                   45379
tagline                  20412
title                    45460
video                    45460
R_average_ratings        45460
V_ratings_count          45460
dtype: int64

In [9]:
# Calculate the mean of the average rating
C = metadata['R_average_ratings'].mean()
print(C)

5.618207215134185


From the output above you can observe that the average rating of a movie on a scale of 10 is around 5.618

In [10]:
# Let's calculate the minimun number of ratings required to be in the chart (M)
M = metadata['V_ratings_count'].quantile(0.90)
print(M)

160.0


M which is 160 is the number of counts a movie has that we will allow/recommend

In [11]:
# Now allow movies that are equal to or greater than 160(M) ratings count
# Keep the relevant movie
q_movies = metadata.copy().loc[metadata['V_ratings_count'] >= M]
q_movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
5  False                                                NaN  60000000   
8  False                                                NaN  35000000   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
4                     [{'id': 35, 'name': 'Comedy'}]   
5  [{'id': 28, 'name': 'Action'}, {'id': 80, 'nam...   
8  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
4                                   NaN  11862  tt0113041                en   
5                                   NaN    949  tt0113277                en   
8                                   NaN   9091  tt0114576                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
4  Father of the Bride Part II   
5                         Heat   
8                 Sudden Death   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943   
1  When siblings Judy and Peter discover an encha...  17.015539   
4  Just when George Banks has recovered from his ...   8.387519   
5  Obsessive master thief, Neil McCauley leads a ...  17.924927   
8  International action superstar Jean Claude Van...    5.23158   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   
5  /zMyfPUelumio3tiDKPffaUpsQTD.jpg   
8  /eoWvKD60lT95Ss1MYNgVExpo5iU.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   
5  [{'name': 'Regency Enterprises', 'id': 508}, {...   
8  [{'name': 'Universal Pictures', 'id': 33}, {'n...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   
5  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
8  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   
5  187436818.0    170.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
8   64350171.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
4  Released  Just When His World Is Back To Normal... He's ...   
5  Released                           A Los Angeles Crime Saga   
8  Released                         Terror goes into overtime.   

                         title  video  R_average_ratings  V_ratings_count  
0                    Toy Story  False                7.7           5415.0  
1                      Jumanji  False                6.9           2413.0  
4  Father of the Bride Part II  False                5.7       

In [12]:
q_movies.shape

(4555, 24)

In [13]:
# Function to compute the weighted rating of each movie

def weighted_ratings(dataframe, M=M, C=C):
    V = dataframe['V_ratings_count']
    R = dataframe['R_average_ratings']
    
    return (V/(V+M) * R) + (M/(M+V) * C)

q_movies['Score'] = q_movies.apply(weighted_ratings, axis=1)
# Show the top movies
q_movies = q_movies.sort_values('Score', ascending=False)
q_movies.reset_index(drop=True, inplace=True)
q_movies.head(5)

adult                              belongs_to_collection     budget  \
0  False                                                NaN   25000000   
1  False  {'id': 230, 'name': 'The Godfather Collection'...    6000000   
2  False                                                NaN   13200000   
3  False  {'id': 263, 'name': 'The Dark Knight Collectio...  185000000   
4  False                                                NaN   63000000   

                                              genres  \
0  [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
1  [{'id': 18, 'name': 'Drama'}, {'id': 80, 'name...   
2  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
3  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...   
4                      [{'id': 18, 'name': 'Drama'}]   

                                       homepage     id    imdb_id  \
0                                           NaN    278  tt0111161   
1                  http://www.thegodfather.com/    238  tt0068646   
2                                           NaN  19404  tt0112870   
3  http://thedarkknight.warnerbros.com/dvdsite/    155  tt0468569   
4    http://www.foxmovies.com/movies/fight-club    550  tt0137523   

  original_language               original_title  \
0                en     The Shawshank Redemption   
1                en                The Godfather   
2                hi  Dilwale Dulhania Le Jayenge   
3                en              The Dark Knight   
4                en                   Fight Club   

                                            overview  popularity  \
0  Framed in the 1940s for the double murder of h...   51.645403   
1  Spanning the years 1945 to 1955, a chronicle o...   41.109264   
2  Raj is a rich, carefree, happy-go-lucky second...   34.457024   
3  Batman raises the stakes in his war on crime. ...  123.167259   
4  A ticking-time-bomb insomniac and a slippery s...   63.869599   

                        poster_path  \
0  /9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg   
1  /rPdtLWNsZmAtoZl9PK7S2wE3qiS.jpg   
2  /2gvbZMtV1Zsl7FedJa5ysbpBx2G.jpg   
3  /1hRoyzDtpgMU7Dz4JF22RANzQO7.jpg   
4  /adw6Lq9FiC9zjYEpOqfq03ituwp.jpg   

                                production_companies  \
0  [{'name': 'Castle Rock Entertainment', 'id': 9...   
1  [{'name': 'Paramount Pictures', 'id': 4}, {'na...   
2           [{'name': 'Yash Raj Films', 'id': 1569}]   
3  [{'name': 'DC Comics', 'id': 429}, {'name': 'L...   
4  [{'name': 'Twentieth Century Fox Film Corporat...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1994-09-23   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1972-03-14   
2            [{'iso_3166_1': 'IN', 'name': 'India'}]   1995-10-20   
3  [{'iso_3166_1': 'GB', 'name': 'United Kingdom'...   2008-07-16   
4  [{'iso_3166_1': 'DE', 'name': 'Germany'}, {'is...   1999-10-15   

        revenue  runtime                                   spoken_languages  \
0  2.834147e+07    142.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  2.450664e+08    175.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2  1.000000e+08    190.0            [{'iso_639_1': 'hi', 'name': 'हिन्दी'}]   
3  1.004558e+09    152.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
4  1.008538e+08    139.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released  Fear can hold you prisoner. Hope can set you f...   
1  Released                         An offer you can't refuse.   
2  Released                               Come... Fall In Love   
3  Released                                    Why So Serious?   
4  Released                            Mischief. Mayhem. Soap.   

                         title  video  R_average_ratings  V_ratings_count  \
0     The Shawshank Redemption  False                8.5           8358.0   
1                The Godfather  False                8.5           6024

In [14]:
recommend_movies = q_movies[['title', 'V_ratings_count', 'R_average_ratings', 'Score']]
recommend_movies.head(22)

title  V_ratings_count  R_average_ratings  \
0          The Shawshank Redemption           8358.0                8.5   
1                     The Godfather           6024.0                8.5   
2       Dilwale Dulhania Le Jayenge            661.0                9.1   
3                   The Dark Knight          12269.0                8.3   
4                        Fight Club           9678.0                8.3   
5                      Pulp Fiction           8670.0                8.3   
6                  Schindler's List           4436.0                8.3   
7                          Whiplash           4376.0                8.3   
8                     Spirited Away           3968.0                8.3   
9                 Life Is Beautiful           3643.0                8.3   
10           The Godfather: Part II           3418.0                8.3   
11  One Flew Over the Cuckoo's Nest           3001.0                8.3   
12                     Forrest Gump           8147.0                8.2   
13          The Empire Strikes Back           5998.0                8.2   
14                           Psycho           2405.0                8.3   
15                 The Intouchables           5410.0                8.2   
16                       Your Name.           1030.0                8.5   
17           Leon: The Professional           4293.0                8.2   
18                   The Green Mile           4166.0                8.2   
19                       GoodFellas           3211.0                8.2   
20               American History X           3120.0                8.2   
21                        Inception          14075.0                8.1   

       Score  
0   8.445869  
1   8.425439  
2   8.421453  
3   8.265477  
4   8.256385  
5   8.251406  
6   8.206639  
7   8.205404  
8   8.196055  
9   8.187171  
10  8.180076  
11  8.164256  
12  8.150272  
13  8.132919  
14  8.132715  
15  8.125837  
16  8.112532  
17  8.107234  
18  8.104511  
19  8.077459  
20  8.074059  
21  8.072105